File: yacc/min_caml_asm/emit.cpp
#include "ast.h"
#include <cassert>
#include <algorithm>
// すでにSaveされた変数の集合 (caml2html: emit_stackset)

static
std::set<std::string> stackset;

// Saveされた変数の、スタックにおける位置 (caml2html: emit_stackmap) *)
static
svec_t stackmap;

static
void save(std::string x) {
	stackset.insert(x);
	svec_t stackmap;
	auto it = std::find(stackmap.begin(), stackmap.end(), x);
	if (it == stackmap.end()) stackmap.push_back(x);
}

static
int locate(std::string x) {
	auto iter = std::find(stackmap.begin(), stackmap.end(), x);
	size_t index = std::distance(stackmap.begin(), iter);
	if (index == stackmap.size()) return -1;
	return index;
}

static
int offset(std::string x) {
	return 4 * locate(x);
}

static
int stacksize() {
	return align(stackmap.size() * 4);
}

static
std::string pp_id_or_imm(Id_or_imm* imm) {
	if (auto im = dynamic_cast<V*> (imm)) return im->v;
	if (auto im = dynamic_cast<C*> (imm))
		return std::string("$") + std::to_string(im->i);
	return std::string("");
}

// 関数呼び出しのために引数を並べ替える(register shuffling) (caml2html: emit_shuffle)

static
std::vector<std::pair<std::string, std::string>>
shuffle(std::string sw, std::vector<std::pair<std::string, std::string>> xys) {
	std::map<std::string, std::string> xys2;
	// remove identical moves
	for (auto xy : xys) {
		if (xy.first == xy.second) continue;
		xys2[xy.first] = xy.second;
	}

	// find acyclic moves
	std::vector<std::pair<std::string, std::string>> xys3;
	std::vector<std::pair<std::string, std::string>> xys4;
	for (auto xy : xys2) {
		auto it = xys2.find(xy.second);
		if (it != xys2.end()) xys3.push_back(xy);
		else xys4.push_back(xy);
	}
	if (xys3.empty() && xys4.empty()) return xys3;
	if (xys4.empty()) {
		auto xy = xys3[0];
		xys3.erase(xys3.begin());
		xys4.push_back(std::pair<std::string, std::string>(xy.second, sw));
		xys4.push_back(xy);
		std::vector<std::pair<std::string, std::string>> xys5;
		for (auto xy2 : xys3) {
			if (xy.second == xy2.first)
				xys5.push_back(std::pair<std::string, std::string>(sw, xy2.second));
			else xys5.push_back(xy2);
		}
		return insert<std::pair<std::string, std::string>>(xys4,shuffle(sw, xys5));
	}
	return insert<std::pair<std::string, std::string>>(xys4,shuffle(sw, xys3));
}

/*
type dest =
  | Tail
  | NonTail of Id.t (* 末尾かどうかを表すデータ型 (caml2html: emit_dest) *)
 */

static
void walk_exp_tail(FILE* oc, Exp* exp);
static
void walk_exp_non_tail(FILE* oc, std::string x, Exp* exp);

// 命令列のアセンブリ生成 (caml2html: emit_g)

static
void walk_e_tail(FILE* oc, E* e) {
	if (auto e_ = dynamic_cast<Ans*> (e))
		return walk_exp_tail(oc, e_->exp.get());
	if (auto e_ = dynamic_cast<Let*> (e)) {
		walk_exp_non_tail(oc, e_->id, e_->exp.get());
		walk_e_tail(oc, e_->e.get());
	}
}

static
void walk_e_non_tail(FILE* oc, std::string x, E* e) {
	if (auto e_ = dynamic_cast<Ans*> (e))
		return walk_exp_non_tail(oc, x, e_->exp.get());
	if (auto e_ = dynamic_cast<Let*> (e)) {
		walk_exp_non_tail(oc, e_->id, e_->exp.get());
		walk_e_non_tail(oc, x, e_->e.get());
	}
}

static
void walk_exp_tail_if(FILE* oc, E* e1, E* e2, std::string b, std::string bn) {
	auto b_else = genid(b + "_else");
	fprintf(oc, "\t%s\t%s\n", bn.c_str(), b_else.c_str());
	auto stackset_back = stackset;
	walk_e_tail(oc, e1);
	fprintf(oc, "%s:\n", b_else.c_str());
	stackset = stackset_back;
	walk_e_tail(oc, e2);
}

static
void walk_exp_non_tail_if(FILE* oc, std::string dest, E* e1, E* e2, std::string b, std::string bn) {
	auto b_else = genid(b + "_else");
	auto b_cont = genid(b + "_cont");
	fprintf(oc, "\t%s\t%s\n", bn.c_str(), b_else.c_str());
	auto stackset_back = stackset;
	walk_e_non_tail(oc, dest, e1);
	auto stackset1 = stackset;
	fprintf(oc, "\tjmp\t%s\n", b_cont.c_str());
	fprintf(oc, "%s:\n", b_else.c_str());
	stackset = stackset_back;
	walk_e_non_tail(oc, dest, e2);
	fprintf(oc, "%s:\n", b_cont.c_str());
	stackset.insert(stackset1.begin(), stackset1.end());
}

static
void walk_exp_args(FILE* oc, std::vector<std::pair<std::string, std::string>>, svec_t);

// 各命令のアセンブリ生成 (caml2html: emit_gprime)

static
void walk_exp_tail(FILE* oc, Exp* exp) {
	// 末尾だったら計算結果を第一レジスタにセットしてret (caml2html: emit_tailret)
	if (dynamic_cast<Nop*> (exp) || dynamic_cast<St*> (exp) || dynamic_cast<Save*> (exp)) {
		auto unit = std::unique_ptr<T>(new Unit());
		walk_exp_non_tail(oc, gentmp(unit.get()), exp);
		fprintf(oc, "\tret\n");
	} else
	if (dynamic_cast<Set*> (exp) || dynamic_cast<SetL*> (exp) || dynamic_cast<Mov*> (exp)
			|| dynamic_cast<Neg*> (exp) || dynamic_cast<Add*> (exp) || dynamic_cast<Sub*> (exp)
			|| dynamic_cast<Ld*> (exp)) {
		walk_exp_non_tail(oc, regs[0], exp);
		fprintf(oc, "\tret\n");
	} else
	if (auto e_ = dynamic_cast<Restore*> (exp)) {
		assert(locate(e_->id) != -1);
		walk_exp_non_tail(oc, regs[0], exp);
		fprintf(oc, "\tret\n");
	} else
	if (auto e_ = dynamic_cast<IfEq*> (exp)) {
		fprintf(oc, "\tcmpl\t%s, %s\n", pp_id_or_imm(e_->imm.get()).c_str(), e_->id.c_str());
		walk_exp_tail_if(oc, e_->e1.get(), e_->e2.get(), "je", "jne");
	} else
	if (auto e_ = dynamic_cast<IfLE*> (exp)) {
		fprintf(oc, "\tcmpl\t%s, %s\n", pp_id_or_imm(e_->imm.get()).c_str(), e_->id.c_str());
		walk_exp_tail_if(oc, e_->e1.get(), e_->e2.get(), "jle", "jg");
	} else
	if (auto e_ = dynamic_cast<IfGE*> (exp)) {
		fprintf(oc, "\tcmpl\t%s, %s\n", pp_id_or_imm(e_->imm.get()).c_str(), e_->id.c_str());
		walk_exp_tail_if(oc, e_->e1.get(), e_->e2.get(), "jge", "jl");
	} else
	if (auto e_ = dynamic_cast<Call*> (exp)) {
		// 関数呼び出しの仮想命令の実装 (caml2html: emit_call)
		walk_exp_args(oc, std::vector<std::pair < std::string, std::string >> (), e_->ids);
		fprintf(oc, "\tjmp\t%s\n", e_->id.c_str());
	}
}

static
void walk_exp_non_tail(FILE* oc, std::string dest, Exp* exp) {
	// 末尾でなかったら計算結果をdestにセット (caml2html: emit_nontail)
	if (dynamic_cast<Nop*> (exp)) {
	} else
	if (auto e_ = dynamic_cast<Set*> (exp)) {
		fprintf(oc, "\tmovl\t$%d, %s\n", e_->v, dest.c_str());
	} else
	if (auto e_ = dynamic_cast<SetL*> (exp)) {
		fprintf(oc, "\tmovl\t$%s, %s\n", e_->l.c_str(), dest.c_str());
	} else
	if (auto e_ = dynamic_cast<Mov*> (exp)) {
		if (dest != e_->id)
			fprintf(oc, "\tmovl\t%s, %s\n", e_->id.c_str(), dest.c_str());
	} else
	if (auto e_ = dynamic_cast<Neg*> (exp)) {
		if (dest != e_->id)
			fprintf(oc, "\tmovl\t%s, %s\n", e_->id.c_str(), dest.c_str());
		fprintf(oc, "\tnegl\t%s\n", dest.c_str());
	} else
	if (auto e_ = dynamic_cast<Add*> (exp)) {
		if (auto v = dynamic_cast<V*> (e_->imm.get())) {
			if (v->v == dest) {
				fprintf(oc, "\taddl\t%s, %s\n", e_->id.c_str(), dest.c_str());
				return;
			}
		}
		if (dest != e_->id)
			fprintf(oc, "\tmovl\t%s, %s\n", e_->id.c_str(), dest.c_str());
		fprintf(oc, "\taddl\t%s, %s\n", pp_id_or_imm(e_->imm.get()).c_str(), dest.c_str());
	} else
	if (auto e_ = dynamic_cast<Sub*> (exp)) {
		if (auto v = dynamic_cast<V*> (e_->imm.get())) {
			if (v->v == dest) {
				fprintf(oc, "\tsubl\t%s, %s\n", e_->id.c_str(), dest.c_str());
				fprintf(oc, "\tnegl\t%s\n", dest.c_str());
				return;
			}
		}
		if (dest != e_->id)
			fprintf(oc, "\tmovl\t%s, %s\n", e_->id.c_str(), dest.c_str());
		fprintf(oc, "\tsubl\t%s, %s\n", pp_id_or_imm(e_->imm.get()).c_str(), dest.c_str());
	} else
	if (auto e_ = dynamic_cast<Ld*> (exp)) {
		if (auto v = dynamic_cast<V*> (e_->imm.get())) {
			fprintf(oc, "\tmovl\t(%s,%s,%d), %s\n", e_->id.c_str(), v->v.c_str(), e_->i, dest.c_str());
			return;
		}
		if (auto c = dynamic_cast<C*> (e_->imm.get())) {
			fprintf(oc, "\tmovl\t%d(%s), %s\n", c->i * e_->i, e_->id.c_str(), dest.c_str());
			return;
		}
		assert(false);
	} else
	if (auto e_ = dynamic_cast<St*> (exp)) {
		if (auto v = dynamic_cast<V*> (e_->imm.get())) {
			fprintf(oc, "\tmovl\t%s, (%s,%s,%d)\n", e_->id.c_str(), e_->id2.c_str(), v->v.c_str(), e_->i);
			return;
		}
		if (auto c = dynamic_cast<C*> (e_->imm.get())) {
			fprintf(oc, "\tmovl\t%s, %d(%s)\n", e_->id.c_str(), c->i * e_->i, e_->id2.c_str());
			return;
		}
		assert(false);
	} else
	// 退避の仮想命令の実装 (caml2html: emit_save)
	if (auto e_ = dynamic_cast<Save*> (exp)) {
		if (allregs.find(e_->id) != allregs.end() && stackset.find(e_->id2) == stackset.end()) {
			save(e_->id2);
			fprintf(oc, "\tmovl\t%s, %d(%s)\n", e_->id.c_str(), offset(e_->id2), reg_sp);
			return;
		}
		assert(stackset.find(e_->id2) != stackset.end());
	} else
	if (auto e_ = dynamic_cast<Restore*> (exp)) {
		// 復帰の仮想命令の実装 (caml2html: emit_restore)
		fprintf(oc, "\tmovl\t%d(%s), %s\n", offset(e_->id), reg_sp, dest.c_str());
	} else
	if (auto e_ = dynamic_cast<IfEq*> (exp)) {
		fprintf(oc, "\tcmpl\t%s, %s\n", pp_id_or_imm(e_->imm.get()).c_str(), e_->id.c_str());
		walk_exp_non_tail_if(oc, dest, e_->e1.get(), e_->e2.get(), "je", "jne");
	} else
	if (auto e_ = dynamic_cast<IfLE*> (exp)) {
		fprintf(oc, "\tcmpl\t%s, %s\n", pp_id_or_imm(e_->imm.get()).c_str(), e_->id.c_str());
		walk_exp_non_tail_if(oc, dest, e_->e1.get(), e_->e2.get(), "jle", "jg");
	} else
	if (auto e_ = dynamic_cast<IfGE*> (exp)) {
		fprintf(oc, "\tcmpl\t%s, %s\n", pp_id_or_imm(e_->imm.get()).c_str(), e_->id.c_str());
		walk_exp_non_tail_if(oc, dest, e_->e1.get(), e_->e2.get(), "jge", "jl");
	} else
	if (auto e_ = dynamic_cast<Call*> (exp)) {
		walk_exp_args(oc, std::vector<std::pair < std::string, std::string >> (), e_->ids);
		auto ss = stacksize();
		if (ss > 0) fprintf(oc, "\taddl\t$%d, %s\n", ss, reg_sp);
		fprintf(oc, "\tcall\t%s\n", e_->id.c_str());
		if (ss > 0) fprintf(oc, "\tsubl\t$%d, %s\n", ss, reg_sp);
		if (allregs.find(dest) != allregs.end() && dest != regs[0]) {
			fprintf(oc, "\tmovl\t%s, %s\n", regs[0].c_str(), dest.c_str());
		}
	} else {
		assert(false);
	}
}

static
void walk_exp_args(FILE* oc, std::vector<std::pair<std::string, std::string>> yrs, svec_t ys) {
	assert(ys.size() <= regs.size() - yrs.size());
	auto sw = std::string() + std::to_string(stacksize()) + ("(" reg_sp ")");
	int i = 0;
	for (auto y : ys) {
		yrs.push_back(std::pair<std::string, std::string>(y, regs[i]));
		i++;
	}
	auto yrs2 = shuffle(sw, yrs);
	for (auto yr : yrs2)
		fprintf(oc, "\tmovl\t%s, %s\n", yr.first.c_str(), yr.second.c_str());
}

static
void walk_fundef(FILE* oc, Fundef* fundef) {
	fprintf(oc, "%s:\n", fundef->name.c_str());
	stackset.clear();
	stackmap.clear();
	walk_e_tail(oc, fundef->body.get());
}

static
void walk_prog(FILE* oc, Prog* prog) {
	fprintf(stderr, "generating assembly...\n");
	fprintf(oc, ".text\n");
	for (auto& fundef : prog->fundefs)
		walk_fundef(oc, fundef.get());
	fprintf(oc, ".globl\tmin_caml_start\n");
	fprintf(oc, "min_caml_start:\n");
	fprintf(oc, ".globl\t_min_caml_start\n");
	fprintf(oc, "_min_caml_start: # for cygwin\n");
	fprintf(oc, "\tpushl\t%%eax\n");
	fprintf(oc, "\tpushl\t%%ebx\n");
	fprintf(oc, "\tpushl\t%%ecx\n");
	fprintf(oc, "\tpushl\t%%edx\n");
	fprintf(oc, "\tpushl\t%%esi\n");
	fprintf(oc, "\tpushl\t%%edi\n");
	fprintf(oc, "\tpushl\t%%ebp\n");
	fprintf(oc, "\tmovl\t32(%%esp),%s\n", reg_sp);
	fprintf(oc, "\tmovl\t36(%%esp),%s\n", regs[0].c_str());
	fprintf(oc, "\tmovl\t%s,%s\n", regs[0].c_str(), reg_hp);
	stackset.clear();
	stackmap.clear();
	walk_e_non_tail(oc, regs[0], prog->e.get());
	fprintf(oc, "\tpopl\t%%ebp\n");
	fprintf(oc, "\tpopl\t%%edi\n");
	fprintf(oc, "\tpopl\t%%esi\n");
	fprintf(oc, "\tpopl\t%%edx\n");
	fprintf(oc, "\tpopl\t%%ecx\n");
	fprintf(oc, "\tpopl\t%%ebx\n");
	fprintf(oc, "\tpopl\t%%eax\n");
	fprintf(oc, "\tret\n");
}

void emit(FILE* oc, Prog* prog) {
	walk_prog(oc, prog);
}

File: yacc/min_caml_asm_rust/src/emit.rs
use ast::*;
use std::iter::Extend;
use std::collections::HashSet;
use std::error::Error;
use std::io::prelude::*;
use std::fs::File;
use std::path::Path;

fn align(i:i32) -> i32 {
  if i % 8 == 0 {
    i
  } else {
    i + 4
  }
}

fn pp_id_or_imm(imm:&IdOrImm) -> String {
  match *imm {
    IdOrImm::V(ref v) => format!("{}", v),
    IdOrImm::C(ref i) => format!("${}", i),
  }
}

// 関数呼び出しのために引数を並べ替える(register shuffling)
fn shuffle(sw:String , xys:&Vec<(String,String)>) -> Vec<(String, String)> {
  // remove identical moves
  let xys:Vec<(String,String)> = xys.iter().filter(|&&(ref x,ref y)|x!=y).cloned().collect();
  // find acyclic moves
  let xys2 = xys.clone();
  let (xys,mut xys_):(Vec<(String,String)>,Vec<(String,String)>) = xys.into_iter().partition(
    |&(_,ref y)|
      xys2.iter().position(|&(ref a, _)| y.eq(a)) != None
    );
  if xys.len() == 0 && xys_.len() == 0 {
    xys
  } else if xys_.len() != 0 {
    xys_.extend(shuffle(sw, &xys));
    xys_
  } else {
    let (ref x,ref y) = xys[0];
    xys_.push((y.clone(),sw.clone()));
    xys_.push((x.clone(),y.clone()));
    let xys = xys[1..].iter().map(|&(ref y_,ref z)|
      if y.eq(y_) {
        (sw.clone(),z.clone())
      } else {
        (y_.clone(),z.clone())
      }
    ).collect();
    xys_.extend(shuffle(sw, &xys));
    xys_
  }
}
pub struct Emit {
  stackset: HashSet<String>,
  stackmap: Vec<String>,
  count: i32,
  file: File,
}
impl Emit {
  pub fn new(file:&str) -> Emit {
    let path = Path::new(&file);
    let file = match File::create(&path) {
      Err(why) => panic!("couldn't create {}: {}", file, why.description()),
      Ok(file) => file
    };
    Emit{
      stackset:HashSet::new(),
      stackmap:vec![],
      count:0,
      file:file,
    }
  }
  fn print(&mut self,s:String) {
    match self.file.write(s.as_bytes()) {
      Err(why) => panic!("couldn't write to {}: {}", s, why.description()),
      Ok(_) => (),
    }
  }
  fn genid(&mut self, x:&String) -> String {
    self.count += 1;
    format!("{}.{}", x, self.count)
  }
  fn gentmp(&mut self, typ:&T) -> String {
    self.count += 1;
    format!("T{}{}", id_of_typ(typ), self.count)
  }
  fn save(&mut self, x:&String) {
    self.stackset.insert(x.clone());
    if self.stackmap.iter().position(|a| x.eq(a)) == None {
      self.stackmap.push(x.clone());
    }
  }
  fn locate(&mut self, x:&String) -> i32 {
    let r = self.stackmap.iter().position(|a| x.eq(a));
    match r {
      None => -1,
      Some(n) => n as i32,
    }
  }
  fn offset(&mut self, x:&String) -> i32 {
    return 4 * self.locate(x);
  }
  fn stacksize(&mut self) -> i32 {
    return align((self.stackmap.len() * 4) as i32);
  }
  // 命令列のアセンブリ生成
  fn walk_e_tail(&mut self, e:&E) {
    match *e {
      E::Ans(ref exp) => self.walk_exp_tail(exp),
      E::Let(ref id, _,ref exp, ref e) => {
        self.walk_exp_non_tail(id, exp);
        self.walk_e_tail(e);
      } 
    }
  }
  fn walk_e_non_tail(&mut self, x:&String, e:&E) {
    match *e {
      E::Ans(ref exp) => self.walk_exp_non_tail(x, exp),
      E::Let(ref id, _, ref exp, ref e) => {
        self.walk_exp_non_tail(id, exp);
        self.walk_e_non_tail(x, e);
      }
    }
  }
  fn walk_exp_tail_if(&mut self, e1:&E, e2:&E, b:&str, bn:&str) {
    let b_else = self.genid(&format!("{}_else", b));
    self.print(format!("\t{}\t{}\n", bn, b_else));
    let stackset_back = self.stackset.clone();
    self.walk_e_tail(e1);
    self.print(format!("{}:\n", b_else));
    self.stackset = stackset_back;
    self.walk_e_tail(e2);
  }
  fn walk_exp_non_tail_if(&mut self, dest:&String, e1:&E, e2:&E, b:&str, bn:&str) {
    let b_else = self.genid(&format!("{}_else", b));
    let b_cont = self.genid(&format!("{}_cont", b));
    self.print(format!("\t{}\t{}\n", bn, b_else));
    let stackset_back = self.stackset.clone();
    self.walk_e_non_tail(dest, e1);
    let stackset1 = self.stackset.clone();
    self.print(format!("\tjmp\t{}\n", b_cont));
    self.print(format!("{}:\n", b_else));
    self.stackset = stackset_back;
    self.walk_e_non_tail(dest, e2);
    self.print(format!("{}:\n", b_cont));
    self.stackset.extend(stackset1)
  }
  // 末尾だったら計算結果を第一レジスタにセットしてret
  fn walk_exp_tail(&mut self, exp:&Exp) {
    match *exp {
      Exp::Nop | Exp::St(_,_,_,_) | Exp::Save(_,_) => {
        let tmp = self.gentmp(&T::Unit);
        self.walk_exp_non_tail(&tmp, exp);
        self.print(format!("\tret\n"))
      }
      Exp::Set(_) | Exp::SetL(_) | Exp::Mov(_) | Exp::Neg(_) | Exp::Add(_,_) |
      Exp::Sub(_,_) | Exp::Ld(_,_,_) => {
        self.walk_exp_non_tail(&format!("{}",REGS[0]), exp);
        self.print(format!("\tret\n"))
      }
      Exp::Restore(ref id) => {
        assert!(self.locate(id) != -1);
        self.walk_exp_non_tail(&format!("{}",REGS[0]), exp);
        self.print(format!("\tret\n"));
      }
      Exp::IfEq(ref id, ref imm, ref e1, ref e2) => {
        self.print(format!("\tcmpl\t{}, {}\n", pp_id_or_imm(imm), id));
        self.walk_exp_tail_if(e1, e2, "je", "jne");
      }
      Exp::IfLE(ref id, ref imm, ref e1, ref e2) => {
        self.print(format!("\tcmpl\t{}, {}\n", pp_id_or_imm(imm), id));
        self.walk_exp_tail_if(e1, e2, "jle", "jg");
      }
      Exp::IfGE(ref id, ref imm, ref e1, ref e2) => {
        self.print(format!("\tcmpl\t{}, {}\n", pp_id_or_imm(imm), id));
        self.walk_exp_tail_if(e1, e2, "jge", "jl");
      }
      Exp::Call(ref id, ref ids) => {
        // 関数呼び出しの仮想命令の実装
        self.walk_exp_args(vec!(), ids);
        self.print(format!("\tjmp\t{}\n", id));
      }
    }
  }
  fn walk_exp_non_tail(&mut self, dest:&String, exp:&Exp) {
    match *exp {
      Exp::Nop => (),
      Exp::Set(v) => self.print(format!("\tmovl\t${}, {}\n", v, dest)),
      Exp::SetL(ref l) => self.print(format!("\tmovl\t${}, {}\n", l, dest)),
      Exp::Mov(ref id) => if dest != id { self.print(format!("\tmovl\t{}, {}\n", id, dest)) },
      Exp::Neg(ref id) => {
        if dest != id {
          self.print(format!("\tmovl\t{}, {}\n", id, dest))
        }
        self.print(format!("\tnegl\t{}\n", dest))
      }
      Exp::Add(ref id, IdOrImm::V(ref v)) if dest.eq(v) =>
        self.print(format!("\taddl\t{}, {}\n", id, dest)),
      Exp::Add(ref id, ref imm) => {
        if dest != id {
          self.print(format!("\tmovl\t{}, {}\n", id, dest))
        }
        self.print(format!("\taddl\t{}, {}\n", pp_id_or_imm(imm), dest))
      }
      Exp::Sub(ref id, IdOrImm::V(ref v)) if dest.eq(v) => {
        self.print(format!("\tsubl\t{}, {}\n", id, dest));
        self.print(format!("\tnegl\t{}\n", dest))
      }
      Exp::Sub(ref id, ref imm) => {
        if dest != id {
          self.print(format!("\tmovl\t{}, {}\n", id, dest))
        }
        self.print(format!("\tsubl\t{}, {}\n", pp_id_or_imm(imm), dest))
      }
      Exp::Ld(ref id, IdOrImm::V(ref v), i) =>
        self.print(format!("\tmovl\t({},{},{}), {}\n", id, v, i, dest)),
      Exp::Ld(ref id, IdOrImm::C(ref j), i) =>
        self.print(format!("\tmovl\t{}({}), {}\n", j * i, id, dest)),
      Exp::St(ref id, ref id2, IdOrImm::V(ref v), i) =>
        self.print(format!("\tmovl\t{}, ({},{},{})\n", id, id2, v, i)),
      Exp::St(ref id, ref id2, IdOrImm::C(ref j), i) =>
        self.print(format!("\tmovl\t{}, {}({})\n", id, j * i, id2)),
      // 退避の仮想命令の実装
      Exp::Save(ref id, ref id2) => {
        if allregs_find(id) && !self.stackset.contains(id2) {
          self.save(id2);
          let offset = self.offset(id2);
          self.print(format!("\tmovl\t{}, {}({})\n", id, offset, REG_SP))
        } else {
          assert!(self.stackset.contains(id2))
        }
      }
      // 復帰の仮想命令の実装
      Exp::Restore(ref id) => {
        let offset = self.offset(id);
        self.print(format!("\tmovl\t{}({}), {}\n", offset, REG_SP, dest))
      }
      Exp::IfEq(ref id, ref imm, ref e1, ref e2) => {
        self.print(format!("\tcmpl\t{}, {}\n", pp_id_or_imm(imm), id));
        self.walk_exp_non_tail_if(dest, e1, e2, "je", "jne");
      }
      Exp::IfLE(ref id, ref imm, ref e1, ref e2) => {
        self.print(format!("\tcmpl\t{}, {}\n", pp_id_or_imm(imm), id));
        self.walk_exp_non_tail_if(dest, e1, e2, "jle", "jg");
      }
      Exp::IfGE(ref id, ref imm, ref e1, ref e2) => {
        self.print(format!("\tcmpl\t{}, {}\n", pp_id_or_imm(imm), id));
        self.walk_exp_non_tail_if(dest, e1, e2, "jge", "jl");
      }
      Exp::Call(ref id, ref ids) => {
        self.walk_exp_args(vec!(), ids);
        let ss = self.stacksize();
        if ss > 0 {
          self.print(format!("\taddl\t${}, {}\n", ss, REG_SP))
        }
        self.print(format!("\tcall\t{}\n", id));
        if ss > 0 {
          self.print(format!("\tsubl\t${}, {}\n", ss, REG_SP))
        }
        if allregs_find(dest) && dest != REGS[0] {
          self.print(format!("\tmovl\t{}, {}\n", REGS[0], dest))
        }
      }
    }
  }
  fn walk_exp_args(&mut self, mut yrs:Vec<(String,String)>, ys:&Vec<String>) {
    assert!(ys.len() <= REGS.len() - yrs.len());
    let sw = format!("{}({})", self.stacksize(), REG_SP);
    let mut i = 0;
    for y in ys.iter() {
      yrs.push((y.clone(), format!("{}",REGS[i])));
      i+=1;
    }
    let yrs2 = shuffle(sw, &yrs);
    for (y,r) in yrs2 {
      self.print(format!("\tmovl\t{}, {}\n", y, r));
    }
  }
  fn walk_fundef(&mut self, fundef:&Fundef) {
    match *fundef {
      Fundef::Fundef(ref name, _, ref body, _) => {
        self.print(format!("{}:\n", name));
        self.stackset.clear();
        self.stackmap.clear();
        self.walk_e_tail(body);
      }
    }
  }
  fn walk_prog(&mut self, prog:&Prog) {
    match *prog {
      Prog::Prog(ref fundefs,ref e) => {
        //fprintf(stderr, "generating assembly...\n");
        self.print(format!(".text\n"));
        for fundef in fundefs {
          self.walk_fundef(fundef)
        }
        self.print(format!(".globl\tmin_caml_start\n"));
        self.print(format!("min_caml_start:\n"));
        self.print(format!(".globl\t_min_caml_start\n"));
        self.print(format!("_min_caml_start: # for cygwin\n"));
        self.print(format!("\tpushl\t%eax\n"));
        self.print(format!("\tpushl\t%ebx\n"));
        self.print(format!("\tpushl\t%ecx\n"));
        self.print(format!("\tpushl\t%edx\n"));
        self.print(format!("\tpushl\t%esi\n"));
        self.print(format!("\tpushl\t%edi\n"));
        self.print(format!("\tpushl\t%ebp\n"));
        self.print(format!("\tmovl\t32(%esp),{}\n", REG_SP));
        self.print(format!("\tmovl\t36(%esp),{}\n", REGS[0]));
        self.print(format!("\tmovl\t{},{}\n", REGS[0], REG_HP));
        self.stackset.clear();
        self.stackmap.clear();
        self.walk_e_non_tail(&format!("{}",REGS[0]), e);
        self.print(format!("\tpopl\t%ebp\n"));
        self.print(format!("\tpopl\t%edi\n"));
        self.print(format!("\tpopl\t%esi\n"));
        self.print(format!("\tpopl\t%edx\n"));
        self.print(format!("\tpopl\t%ecx\n"));
        self.print(format!("\tpopl\t%ebx\n"));
        self.print(format!("\tpopl\t%eax\n"));
        self.print(format!("\tret\n"));
      }
    }
  }
}
pub fn emit(file:&String, prog:&Prog) {
  Emit::new(file).walk_prog(prog)
}
